AK: Fix a race condition with WeakPtr<T>::strong_ref and destruction

tomuta opened the pull request:
Since RefPtr<T> decrements the ref counter to 0 and after that starts
destructing the object, there is a window where the ref count is 0
and the weak references have not been revoked.

Also change WeakLink to be able to obtain a strong reference
concurrently and block revoking instead, which should happen a lot
less often.

Fixes #4621

@bcoles Does this solve this issue for you? I haven't tested it a whole lot, but I also haven't run into the crash you're seeing.
@bugaevc Memory order review, please? ;-)

bcoles commented:
> @bcoles Does this solve this issue for you? I haven't tested it a whole lot, but I also haven't run into the crash you're seeing.

I'll let you know as soon as I figure out which one of the commits in the last 12 hours broke my build lol.

bcoles commented:
> 
> 
> > @bcoles Does this solve this issue for you? I haven't tested it a whole lot, but I also haven't run into the crash you're seeing.
> 
> I'll let you know as soon as I figure out which one of the commits in the last 12 hours broke my build lol.

Some stats:

Before I created issue #4621, I booted 20 times and got 2 crashes (on latest `master` as of 13 hours ago).

Before testing this commit, I booted 3 times and got 1 crash (on latest `master`).

Testing with this commit, I booted 20 times and got 2 crashes.

Both crashes were in `SharedInodeVMObject`, rather than due to a invalid reference count, which I think is an improvement (?)

While not an accurate or conclusive test (this bug, or a similar bug, has existed for a long time), this is a strong indication that the reference count portion of the issue has been resolved.

```
[Taskbar(11:11)]: ASSERTION FAILED: inode().is_shared_vmobject(*this)
../Kernel/VM/SharedInodeVMObject.cpp:61 in virtual Kernel::SharedInodeVMObject::~SharedInodeVMObject()
[Taskbar(11:11)]: 0xc0119076  __assertion_failed(char const*, char const*, unsigned int, char const*) +185
[Taskbar(11:11)]: 0xc01b5496  Kernel::SharedInodeVMObject::~SharedInodeVMObject() +80
[Taskbar(11:11)]: 0xc01b54c6  Kernel::SharedInodeVMObject::~SharedInodeVMObject() +28
[Taskbar(11:11)]: 0xc01b3a88  Kernel::Region::~Region() +280
[Taskbar(11:11)]: 0xc01d9860  AK::OwnPtr<Kernel::Region>::clear() +38
[Taskbar(11:11)]: 0xc01762f0  Kernel::Process::deallocate_region(Kernel::Region&) +386
[Taskbar(11:11)]: 0xc0189816  Kernel::Process::sys$munmap(void*, unsigned long) +234
[Taskbar(11:11)]: 0xc017da16  syscall_handler +1710
[Taskbar(11:11)]: 0xc017d360  syscall_asm_entry +49
```

```
[Audio.MenuApplet(21:21)]: ASSERTION FAILED: inode().is_shared_vmobject(*this)
../Kernel/VM/SharedInodeVMObject.cpp:61 in virtual Kernel::SharedInodeVMObject::~SharedInodeVMObject()
[Audio.MenuApplet(21:21)]: 0xc0119076  __assertion_failed(char const*, char const*, unsigned int, char const*) +185
[Audio.MenuApplet(21:21)]: 0xc01b5496  Kernel::SharedInodeVMObject::~SharedInodeVMObject() +80
[Audio.MenuApplet(21:21)]: 0xc01b54c6  Kernel::SharedInodeVMObject::~SharedInodeVMObject() +28
[Audio.MenuApplet(21:21)]: 0xc01b3a88  Kernel::Region::~Region() +280
[Audio.MenuApplet(21:21)]: 0xc01d9860  AK::OwnPtr<Kernel::Region>::clear() +38
[Audio.MenuApplet(21:21)]: 0xc01762f0  Kernel::Process::deallocate_region(Kernel::Region&) +386
[Audio.MenuApplet(21:21)]: 0xc0189bcf  Kernel::Process::sys$mremap(AK::Userspace<Kernel::Syscall::SC_mremap_params const*>) +619
[Audio.MenuApplet(21:21)]: 0xc017da16  syscall_handler +1710
[Audio.MenuApplet(21:21)]: 0xc017d360  syscall_asm_entry +49
```

bcoles commented:
I've been using this patch in my local branch while doing development. A few more crashes for `SharedInodeVMObject`, but none for reference counters.

However, I'm fairly certain I've never seen this bug before, and it is 100% reproducible by closing the Network notification popup that appears on system startup. Presumably it was introduced by this patch.

```
USERSPACE(26) ASSERTION FAILED: !(current_consumers & 1u)
.././AK/Weakable.h:97
[NotificationServer(26:26)]: Terminating NotificationServer(26) due to signal 6
[FinalizerTask(3:3)]: 0xdeadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) +181
0xdeadc0de  Kernel::Scheduler::context_switch(Kernel::Thread*) +155
0xdeadc0de  Kernel::Scheduler::pick_next() +956
0xdeadc0de  Kernel::Processor::check_invoke_scheduler() +125
0xdeadc0de  Kernel::Thread::die_if_needed() +163
0xdeadc0de  syscall_handler +2181
0xdeadc0de  syscall_asm_entry +49
0x02f7d5f0
0x02f757e9
0x02f760b3
0x02ffe264
0x02ffdf8c
0x032714b7
0x032714de
0x01e55577
0x01e55604
0x1019c02c

```

bcoles commented:
Actually, that crash appears to be reproducible by closing any application, not just the Network notification popup.

```
USERSPACE(28) ASSERTION FAILED: !(current_consumers & 1u)
.././AK/Weakable.h:97
[TextEditor(28:28)]: Terminating TextEditor(28) due to signal 6
[FinalizerTask(3:3)]: 0xdeadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) +181
0xdeadc0de  Kernel::Scheduler::context_switch(Kernel::Thread*) +155
0xdeadc0de  Kernel::Scheduler::pick_next() +956
0xdeadc0de  Kernel::Processor::check_invoke_scheduler() +125
0xdeadc0de  Kernel::Thread::die_if_needed() +163
0xdeadc0de  syscall_handler +2181
0xdeadc0de  syscall_asm_entry +49
0x02ced5f0
0x02ce57e9
0x02ce60b3
0x02d6e264
0x02d6df8c
0x02ffc4b7
0x02ffc4de
0x033d88fb
0x033d893a
0x14a8f02c
```

tomuta commented:
Yeah that's a new assertion added by this PR. Looks like `WeakLink::revoke` was called twice, or on a bogus pointer...

bcoles commented:
> 
> 
> Yeah that's a new assertion added by this PR. Looks like `WeakLink::revoke` was called twice, or on a bogus pointer...

All apps are triggering it lol. This results in a coredump, which triggers `CrashReporter`. When `CrashReporter` is closed, it crashes, resulting in a crash report for `CrashReporter`, which ...

![I heard you like crashes so I crashed CrashReporter so you can crash while reporting crashes](https://user-images.githubusercontent.com/434827/103319579-6970b800-4a86-11eb-8d79-55e13ff4b5ba.png)

tomuta commented:
Well I guess I really should have tested it better ;) But again, I didn't test it much ;) Interestingly, for me the crash reporter doesn't show up...

tomuta commented:
This is probably the line to blame:
https://github.com/SerenityOS/serenity/blob/91b65ec328d30d7aa1c84597ccf831fc13faf221/Libraries/LibGUI/Application.cpp#L107
I assumed that it can only be called once...

bcoles commented:
> 
> 
> Well I guess I really should have tested it better ;) But again, I didn't test it much ;) Interestingly, for me the crash reporter doesn't show up...

I'm testing on `master`, but with your two patched files pulled in. I believe the `CrashReporter` was merged after you branched this PR from `master`.

tomuta commented:
> > Well I guess I really should have tested it better ;) But again, I didn't test it much ;) Interestingly, for me the crash reporter doesn't show up...
> 
> I'm testing on `master`, but with your two patched files pulled in. I believe the `CrashReporter` was merged after you branched this PR from `master`.

I had rebased my branch on top of master, but maybe I wasn't patient enough. I removed the call from `~Application` and that fixes this assertion.

tomuta commented:
hm seems to cause some hangs, but now that I don't see symbolified stack traces in SystemMonitor anymore it's hard to tell what exactly or why...

awesomekling commented:
> hm seems to cause some hangs, but now that I don't see symbolified stack traces in SystemMonitor anymore it's hard to tell what exactly or why...

Maybe you can use the Profiler to see what they're doing :)

bcoles commented:
> 
> 
> hm seems to cause some hangs, but now that I don't see symbolified stack traces in SystemMonitor anymore it's hard to tell what exactly or why...

Applications no longer crash upon closing. But after the window disappears, they're still running, and chewing up 100% CPU.

![100% CPU uage](https://user-images.githubusercontent.com/434827/103321193-8f995680-4a8c-11eb-936f-d559b46ffc2a.png)

tomuta commented:
> > hm seems to cause some hangs, but now that I don't see symbolified stack traces in SystemMonitor anymore it's hard to tell what exactly or why...
> 
> Maybe you can use the Profiler to see what they're doing :)

Hm not getting too many clues out of it using the Profiler :( Not sure where exactly it's spinning

awesomekling commented:
> Hm not getting too many clues out of it using the Profiler :( Not sure where exactly it's spinning

Here's a total hunch: maybe it's spinning trying to lock the bottom bit of a "locked" RefPtr that's actually somewhere in freed memory?

tomuta commented:
> > Hm not getting too many clues out of it using the Profiler :( Not sure where exactly it's spinning
> 
> Here's a total hunch: maybe it's spinning trying to lock the bottom bit of a "locked" RefPtr that's actually somewhere in freed memory?

Yeah the `clear` function in there showed up, but it wasn't high up in the usage

tomuta commented:
Interesting, the spinning when applications exit is because I removed the call to `revoke_weak_ptrs` from `~Application`. When I put it back into place (and fix it up so that it can handle multiple calls) then no spinning.

tomuta commented:
Still haven't been able to figure out why removing the call to `revoke_weak_ptrs` from `~Application` causes stuff to spin, but I updated `revoke_weak_ptrs` to be able to be called more than once for now...

awesomekling commented:
Maybe something is doing `Application::the()->whatever` after `Application` is destroyed?

tomuta commented:
> Maybe something is doing `Application::the()->whatever` after `Application` is destroyed?

Yes, and the problem turns out to be the `Application::the` implementation, which only works in user mode and implicitly calls `WeakPtr<T>::unsafe_ptr` which is...well...unsafe. Changing the implementation to `return s_the->strong_ref().ptr();` works as intended and returns `nullptr` as soon as the `Application`'s ref count drops to 0...

tomuta commented:
I guess enhancing `WeakLink::unsafe_ptr` to do a similar (though unsafe) ref count check could probably solve it, too. I'll give that a shot...

tomuta commented:
Well that `WeakLink::unsafe_ptr` enhancement turns out to be.... difficult. That's because WeakPtr<T> doesn't always have the full type definition available when pulling in that code, so I can't check if it's a `RefCounted` based type :(

awesomekling commented:
> Well that `WeakLink::unsafe_ptr` enhancement turns out to be.... difficult. That's because WeakPtr doesn't always have the full type definition available when pulling in that code, so I can't check if it's a `RefCounted` based type :(

Seems fine to just make a strong ref everywhere though?

tomuta commented:
Do you mean change `WeakPtr::operator->` and `WeakPtr::operator*` to return a `RefPtr<T>`?

tomuta commented:
But then you'd have to do something like `(*my_weak_ptr)->foo()`...

tomuta commented:
Or just get rid of `operator->` and `operator*`? But there's a ton of user mode code that uses it all over the place...

tomuta commented:
Luckily we don't allow it in kernel code...

awesomekling commented:
> Or just get rid of `operator->` and `operator*`? But there's a ton of user mode code that uses it all over the place...

Yeah this would be painful but maybe it’s worth it. The current usage pattern is not thread friendly at all.

tomuta commented:
True, that's why we disallow it in the kernel. But I think for now I just added a comment to `Application::the`, at least that explains the weird behavior I was tracking down.

tomuta commented:
Or, I guess I could make the `WeakPtr::unsafe_ptr` call `strong_ref().ptr()`...
